### PR TITLE
Initial fix for hanging bug when NaN/inf loss is encountered

### PR DESCRIPTION
**Pull request by redoctopus**
As seen in #532 and some other experiments I've run.

Signed-off-by: Jocelyn Huang <jocelynh@nvidia.com>

### CODE REVIEW

File: nemo/backends/pytorch/actions.py [hunk 1]
```diff
@@ -1356,29 +1369,25 @@ def train(
...
-                        nan = True
-                        break
+                        if (
+                            (stop_on_nan_loss)
+                            or (self._optim_level not in AmpOptimizations)
+                            or (self._optim_level == Optimization.mxprO0)
+                        ):
+                            # Set flag here and terminate at next eval step (to avoid desync and hanging).
+                            self.nan_or_inf = True
+                            logging.warning('Loss is NaN or inf, will terminate at next eval step')
```
Comment (blisc) [inline nemo/backends/pytorch/actions.py:1384]: Can you also print the step here?
Something like ('Loss is NaN or inf at %d, will terminate at next eval step', self.step)

File: nemo/core/neural_factory.py [hunk 2]
```diff
@@ -135,9 +135,9 @@ def train(
                 "algorithmic" batch size per GPU/worker = batches_per_step*
                 batch_size
             stop_on_nan_loss: (default: False) If set to True, the training
-                will stop if loss=nan. If set to False, the training will
-                continue, but the gradients will be zeroed before next
-                mini-batch.
+                will stop if loss=nan or inf. If set to False, the training
+                will continue. Note that if apex is not used, or if
```
Comment (blisc) [inline nemo/core/neural_factory.py:139]: change apex to amp or apex.amp

File: nemo/backends/pytorch/actions.py (hunk 1 above)
Comment (redoctopus) [reply nemo/backends/pytorch/actions.py]: Done

File: nemo/core/neural_factory.py (hunk 2 above)
Comment (redoctopus) [reply nemo/core/neural_factory.py]: Done

File: nemo/backends/pytorch/actions.py [hunk 3]
```diff
@@ -515,6 +516,18 @@ def _eval(self, tensors_2_evaluate, callback, step, verbose=False):
                 #         self.local_rank, world_size
                 #     )
                 # )
+
+                # Check if we should terminate due to NaN/inf on any workers
+                # We use dtype=int because nccl backend doesn't support torch.bool
+                nan_inf_tensor = torch.tensor(self.nan_or_inf, dtype=int).cuda()
```
Comment (okuchaiev) [inline nemo/backends/pytorch/actions.py:522]: this will fail on CPU. Can we check if we are on GPU before doing this?

Comment (blisc) [reply nemo/backends/pytorch/actions.py:522]: Perhaps we should just do a check for distributed. Since this should not be a bug if we are running on a single process, and I do not believe we support distributed CPU processes.

Comment (redoctopus) [reply nemo/backends/pytorch/actions.py:522]: This is wrapped within `if dl_nm.placement == DeviceType.AllGpu`; that should take care of this problem, right?
Good catch that it will not stop if there is only a training loop. I did eval to avoid another sync on each step, or a sync on every N steps, but the latter may be a more robust solution. What do you guys think?
